Add Nimbella CLI tests to public repo.

jthomas opened the pull request:
Set up bats test framework to run shell-script based tests. Started ported internal tests to public repo. Re-wrote tests to ensure no side-effects.

This is a WIP for the addition of functional tests to the CLI repo.

rabbah commented on tests/config_types/test.bats:
```diff
@@ -0,0 +1,47 @@
+#
```
This repo is Apache licensed.

jthomas replied:
Aw - good catch! Will sort out all the license headers.

jthomas commented:
Work on this is still on-going (see latest commit). It is fairly straightfoward (if slightly laborious). I've found a few issues with tests failing with the latest master version of the public CLI versus the published CLI. Will open issues once I've finished this up. Hope to have a first pass finished this week.

jthomas commented:
I'm probably 90% of the way through the tests now....

jthomas commented:
I've now completed porting all the tests that don't require a project apart from the following.

- CommanderUsage - I need to think how to port this test and whether it's better situated in the commander repo.
- LoginAuth - Same as above - need to consider how to port this.

@joshuaauerbachwatson @rabbah Might be worth having a glance over the PR to spot any obvious issues. Each test scripts and related files are grouped into individual directories. 

It takes  ~9 minutes to run the whole test suite (46 tests) on my local development machine.

I've discovered a few issues whilst running the tests that I'm still hunting down (a few of them fail). Once I've worked out where the issue is - I'll update the PR and it'll be ready for merge.

jthomas commented:
I've fixed a few small things in the tests discovered whilst running the entire test suite - just need to finish off investigating the sporadic remote builder test errors.

The bats framework has the option to run tests in parallel using the following command:
```
bats -j 4  --no-parallelize-within-files -r */*bats
```

I wrote the test files to assume tests within them run in sequential order - so we can parallelise across files fine but not within them. **Testing this out reduces the test suite run time to under 3 minutes (using 4 parallel jobs on my MBP).** There is one extra test error which occurs I'll have a look at but I'm impressed it worked so well.

jthomas commented:
All tests are finally passing as expected 🕺🕺🕺. This is ready for review & merging. 

```
$ bats -r */*bats                                                                                          
 ✓ can retrieve web actions using action wrapping
 ✓ cannot deploy project with invalid annotations
 ✓ cannot deploy project with invalid env
 ✓ cannot deploy project with invalid parameters
 ✓ cannot deploy project with invalid top-level parameter
 ✓ deploying project with nothing to deploy
 ✓ deploying project with empty zip file should fail
 ✓ deploying project using '.env' default
 ✓ deploying project using alternative file 'test.env'
 ✓ deploying project using an environment variable
 ✓ deploying project using clashing resource identifiers
 ✓ deploying project using misspelled actions
 ✓ should not deploy filtered web resources
 ✓ should not deploy filtered package resources
 ✓ deploy project whilst ignoring local files
 ✓ deploy project including web content
 ✓ deploy project excluding folder
 ✓ deploy project including folder
 ✓ deploy project with multiple include/excludes
 ✓ deploy project incrementally with no changes
 ✓ deploy project incrementally with action changes
 ✓ deploy project incrementally with metadata & web changes
 ✓ deploy project incrementally with include changes
 ✓ deploy project with multi-level parameter substitution
 ✓ deploy project with production namespace conflict
 ✓ deploy project with test namespace conflict
 ✓ deploy project after freeing namespace
 ✓ deploy go projects with remote build
 ✓ invoke remotely built go lang actions
 ✓ deploy java projects with remote build
 ✓ invoke remotely built java lang actions
 ✓ deploy php projects with remote build
 ✓ invoke remotely built php lang actions
 ✓ deploy python projects with remote build
 ✓ invoke remotely built python lang actions
 ✓ deploy swift projects with remote build
 ✓ invoke remotely built swift lang actions
 ✓ can execute actions using dotnet runtime
 ✓ invoking sequence runs individual actions
 ✓ deploy projects with shared build
 ✓ actions in projects with shared build work
 ✓ deploy project with split-level parameters in config
 ✓ deploy project with default web values
 ✓ deploy project with invalid runtime
 ✓ watch for project changes that trigger a build
 ✓ watch for project changes that do not trigger a build
 ✓ deploy project web resources with caching on
 ✓ deploy project web resources with caching off
 ✓ deploy project with web content included
 ✓ deploy project with config containing YAML aliases

50 tests, 0 failures
```

rabbah commented on tests/shared_build/build/package.json:
```diff
@@ -0,0 +1,14 @@
+  "main": "index.js",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/nimbella-corp/workbench.git"
```
can we remove these references?

rabbah commented on tests/watch/expected-watcher.output:
```diff
@@ -0,0 +1,23 @@
+Deployment complete.  Resuming watch.
+
+
+Deploying '/Users/jamesthomas/code/nimbella/nimbella-cli/tests/watch/watched' due to change in '/Users/jamesthomas/code/nimbella/nimbella-cli/tests/watch/watched/packages/default/hello2/index.js'
```
Do the tests work around `/Users/jamesthomas`?

jthomas replied:
Good spot - nope! Will work out how to fix this...

rabbah reviewed (commented):
LGTM.

jthomas commented:
> Beginning my review by installing bats, applying the PR locally, and running the tests. As communicated on slack, I think there's an error in an earlier PR that made 30 of the tests fail (I will just fix that ASAP and separately). WIth that fix in place it reports 4 failures and 3 "not run." I can't account for the "not run" cases ... perhaps you can tell me how to track those down. Here are the failures:
> 
> ```
>  ✗ deploy project with multiple include/excludes
>    (from function `setup_file' in test file incremental/test.bats, line 5)
>      `cp -r $BATS_TEST_DIRNAME/test-resources/example/* $BATS_TEST_DIRNAME/test-project' failed with status 64
>    usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file
>           cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory
> ```
> 
> Not sure what's going on there.

This was an issue in the test (missing empty directory in git repo) - I've now pushed a fix. 

> ```
>  ✗ begin 29 invoke remotely built go lang actions
> ```
> 
> Timed out after 3 seconds. You should increase the timeout in the `project.yml` rather than worrying about whether 3 seconds is reasonable. I've had to do that for many other actions in other tests.

Can you tell me all the tests failed on the timeouts and I'll increase them?

> ```
>  ✗ begin 38 can execute actions using dotnet runtime
>    (from function `assert_output' in file /usr/local/lib/bats-assert/src/assert.bash, line 231,
>     in test file runtime-dotnet/test.bats, line 15)
>      `assert_output --partial '"greeting": "Hello, stranger!"'' failed
>    
>    -- output does not contain substring --
>    substring (1 lines):
>      "greeting": "Hello, stranger!"
>    output (3 lines):
>      {
>        "error": "code must be binary (zip file)."
>      }
> ```
> 
> I'm not sure what is going there. We may need to get to the bottom of it jointly.

Hmm this doesn't look good. If you run the nim commands in the test manually - do they work?

```
nim project deploy runtime-dotnet/
nim  action invoke test_dotnet_runtime/hello

> ```
>  ✗ begin 45 watch for project changes that trigger a build
> ...
>      < Watching '/Users/joshuaauerbach/nimbella/public/nimbella-cli/tests/watch/watched' [use Control-C to terminate]
>      ---
>      > Watching '/Users/jamesthomas/code/nimbella/nimbella-cli/tests/watch/watched' [use Control-C to terminate]
> ...
> ```
> 
> I omitted the rest of the output of the test. It looks like you bake in a local file in the comparison transcript. I can see how this can happen. In the original test, because the deployer echos what it is given as an argument, it only echoed a relative path. Can you make that happen with this test?

This should be now fixed. I've tried not to use relative paths in the tests to ensure the test files can be executed from any directory in the repo. I've fixed the `diff` to use a template source file to fix this. Bats gives you a $BATS_TEST_DIRNAME variable in the tests to refer to the current directory in which the test file lives. Can you re-test?

> 
> Apart from these problems with individual tests (and the mysterious missing 3) I think the README needs to warn the user about possible consequences of running the tests. I see that the tests depend on package isolation and tests cleaning up after themselves by deleting packages. This avoids the problem in the original tests where we were constantly cleaning namespaces in their entirety. But, it's not completely safe because the package name chosen by the test can collide with a package name in use in the namespace for other reasons.

Good point - will do.

joshuaauerbachwatson commented:
I tried again after changes.   Now I get 48 out of 50 tests succeeding, the mysterious omissions are gone. 

The remaining failures are

1.  The remote go invoke phase still fails with a timeout.   I edited the `project.yml` to increase the timeout to 10000 but I am still getting a timeout during initialization.   This is very suspicious.  I don't think it's just a timeout; rather, something weird is going on.
2. The dotnet test still shows the same error as before.   The corresponding test in the old format works.

I understand you didn't do anything to fix those particular tests, so this is not surprising.  @jthomas please advise whether we should just merge this (ok with me; the addition of these tests does nothing much to break anything until we start running them routinely).   In that case, let's open two narrow issues to work on the two remaining failures.  I'll help.  I'll start now by probing what's different about the old and new versions of the dotnet test.

joshuaauerbachwatson commented:
> 2\. The dotnet test still shows the same error as before.   The corresponding test in the old format works.

Probably false alarm.  The test employs a pre-built zip file (binary).  Since I tested the PR by patching in the diff, it didn't handle the binary file correctly and it ended up empty.   I will assume that file is actually committed correctly in the PR (especially if it works for you).   So, I should just check again after doing the merge.

joshuaauerbachwatson commented:
The issue with the go remote build test is attributable to the system under test, not the test itself.   If I run

```
bats remote-build-go/*bats
```
(testing the system as built in the public repo) it fails.   But, if I run

```
NIM=nim bats remote-build-go/*bats
```
(testing my globally installed `nim`, which was built using the standard build running in the private repo) it works.

Since both builds use identical code, the difference has to be due to some differences in incorporated meta-data.   I strongly suspect `runtimes.json`, which is indeed handled differently in public vs private.

Anyway, I can now say there are no known problems with this PR.   There's a _possible_ problem with the zipfile in dotnet but that is almost certainly my error in how I applied the PR.   There is no problem with the remote go test, just a likely problem elsewhere in the public repo that needs to be fixed.   I'm approving it.

joshuaauerbachwatson reviewed (approved):
I'm approving based on now being convinced that all 50 tests do work.   I'm not inspecting every test at this time to see whether it really tests what its counterpart tested ... I have reasonable faith in that, and we should move forward with changing our internal testing to use these tests instead of their counterparts.

jthomas commented:
Great - we can move forward with this. @bhageena is also going to sanity check the tests by running them locally to check on Linux (rather than OS X). Assuming he doesn't find anything major - we'll merge. I've opened a small issue based on his feedback already.
